internal/runners/secrets/secrets.go

```diff
@@ -139,25 +140,38 @@ func definedSecrets(proj *project.Project, secCli *secretsapi.Client, cfg keypai
 	}
 
 	if filter != "" {
-		secretDefs = filterSecrets(proj, cfg, auth, secretDefs, filter)
+		secretDefs, err = filterSecrets(proj, cfg, auth, secretDefs, filter)
+		if err != nil {
+			return nil, errs.Wrap(err, "Could not filter secrets")
+		}
 	}
 
 	return secretDefs, nil
 }
 
-func filterSecrets(proj *project.Project, cfg keypairs.Configurable, auth *authentication.Auth, secrectDefs []*secretsModels.SecretDefinition, filter string) []*secretsModels.SecretDefinition {
+func filterSecrets(proj *project.Project, cfg keypairs.Configurable, auth *authentication.Auth, secrectDefs []*secretsModels.SecretDefinition, filter string) (defs []*secretsModels.SecretDefinition, rerr error) {
 	secrectDefsFiltered := []*secretsModels.SecretDefinition{}
 
 	oldExpander := project.RegisteredExpander("secrets")
 	if oldExpander != nil {
-		defer project.RegisterExpander("secrets", oldExpander)
+		defer func() {
+			rerr = project.RegisterExpander("secrets", oldExpander)
```

Comment:
I just realized that maybe this is overwriting any of the `return nil, errs.Wrap(...)` lines below? Maybe we need to have 

```
if rerr != nil {
  rerr = project.RegisterExpander(...)
}
```

Reply:
That's a good point! However, I've gone with a little different approach as I want to ensure the defer is always executing the function.
```
defer func() {
	err := project.RegisterExpander("secrets", oldExpander)
	if err != nil {
		rerr = errs.Wrap(err, "Could not register old secrets expander")
	}
}()
```

Reply:
That's fair, but if there's an error in `RegisterExpander`, it will overwrite any error in the function. Maybe `if err != nil && rerr == nil`? I honestly don't know how else we'd handle this case of two errors happening. I think we should return the first one that occurred.

Reply:
In this case the error is overwritten, but wrapped. Only localized errors will be reported to users and the log will contain the full error chain. I think this is okay but if I'm missing your point please let me know.

Reply:
Let us suppose line 169 fails:

```
return nil, errs.Wrap(err, "Could not expand filter")
```

Now the defer triggers and `project.RegisterExpander()` fails.

`rerr`, which previously contained a wrapped "Could not expand filter", will now discard its current value in favor of a wrapped "Could not register old secrets expander". Thus the original error in the function is lost.

Unless I'm missing the point?

Reply:
No, you're right I was missing the error target to `errs.Wrap`. I looked through our `errs` package and found `errs.Pack` which is being used here: https://github.com/ActiveState/cli/blob/master/internal/runbits/runtime/rationalize.go#L30

Take a look at the latest commit and let me know what you think.